src/main/java/pl/confitura/jelatyna/registration/voucher/VoucherRepository.java

```diff
@@ -0,0 +1,29 @@
+package pl.confitura.jelatyna.registration.voucher;
+
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.Repository;
+import org.springframework.data.rest.core.annotation.RepositoryRestResource;
+import org.springframework.data.rest.core.annotation.RestResource;
+import org.springframework.security.access.prepost.PreAuthorize;
+
+import java.util.List;
+
+
+@RepositoryRestResource(path = "vouchers")
+interface VoucherRepository extends Repository<Voucher, String> {
+
+    @RestResource(exported = false)
+    @PreAuthorize("@security.isAdmin()")
+    Voucher save(Voucher token);
+
+    @PreAuthorize("@security.isAdmin()")
+    List<Voucher> findAll();
+
+    @RestResource(exported = false)
+    boolean existsById(String token);
+
+    @Query("select t from Voucher t where t not in (select p.voucher from pl.confitura.jelatyna.registration.Participant p)")
```

Comment:
lets break it into separate lines to make it slightly more readable.
I would use different variable. I am used to use variables that have something in common with the entity. There is no `t` in `Voucher` :) is that for a `table` ? But Voucher is not a table - it is an entity. I would use `v`. If this is reserved for a `view` or something like that then `vo`

I also like writting SQL keywords with Uppercase, so that you can easily  see where is SQL, and where entites.

Reply:
i was refactoring it few times. I started it as `Token` then changed to `Ticket` and finally decided to leave it as `Voucher`

Reply:
:)